oreo.go

```diff
@@ -378,130 +379,130 @@ func (c *Client) Do(req *http.Request) (resp *http.Response, err error) {
 	return resp, err
 }
 
-func (c *Client) Get(urlStr string) (resp *http.Response, err error) {
+func (c *Client) Get(ctx context.Context, urlStr string) (resp *http.Response, err error) {
```

Comment:
I dont think I want to break these apis.  What if we added a `WithContext` to the `Client` object?

something like:
```
func (c *Client) WithContext( ctx context.Context ) *Client {
  // all the With functions return a copy of the client, they dont
  // directly modify the client object.
  cp := *c
  cp.ctx = ctx
  return &cp
}
```
This would allow us to do things like this in `Get`:
```
req := RequestBuilder(parsed).WithContext(c.ctx).WithMethod("GET").Build()
```

Users would add new contexts like:
```
o := oreo.New()
o.WithContext( context.Background() ).Get(...)
```
This will leave the original `o` alone, so that later calls to `o.Get(...)` will be without a context.

Reply:
Yes, I feel like contexts should be optional, just like they are optional in the Go http library.

Reply:
How is it not goroutine safe?  o.WithContext() returns a copy.

Reply:
My bad, it does return a copy. Though that copy implementation may not be correct since its a shallow one. They may share the same `preCallbacks` slice?

Reply:
yes, that is by design, if a user doesn't want previously registered callbacks, they would call:
```
o.WIthoutCallbacks().WithContext(ctx).Get(...)
```

Reply:
My bad :) you are right, this is not ideal for callbacks, we should probably just implement a `copy` routine to dup the callback slice, so that this does not call problems:
```
cp := o.WithCallback(...)
cp.WithCallback(...)
```
we obviously dont want `cp.WithCallback(...)` to modify the `o` object, which I think is what it doing now.

Reply:
But consider this scenario:
```go
someOreo := oreo.New()

otherOreo := someOreo.WithContext(ctx)
otherOreo.WithPreCallback(otherCallback)

someOreo.Get(...) // Triggers otherCallback
```

Reply:
This test fails:
```go
func TestOreoWithPreCallbackCopy(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_, ok := r.Header["Authorization"]
		assert.True(t, ok)
		fmt.Fprintf(w, "OK")
	}))
	defer ts.Close()

	callback := func(req *http.Request) (*http.Request, error) {
		req.SetBasicAuth("user", "pass")
		return req, nil
	}

	c := New()
	other := c.WithPreCallback(callback)

	resp, err := c.Get(context.Background(), ts.URL)
	assert.Nil(t, resp)
	assert.Error(t, err)

	resp, err = other.Get(context.Background(), ts.URL)
	assert.NotNil(t, resp)
	assert.Nil(t, err)
}
```

Reply:
Hmm, I think this actually works as designed afterall:
https://github.com/coryb/oreo/blob/master/oreo.go#L137

we append to the copy when calling `With*Callback`, which is not mutating the source.